### PR TITLE
Magpietts 2503 po july2025

### PR DESCRIPTION
**Pull request by shehzeen**
Preference optimization updates. 
1. Added thresholding for CERs to ignore groups with very high CERs
2. Local transformer GRPO enabled

### CODE REVIEW

File: nemo/collections/tts/models/magpietts_preference_optimization.py [hunk 1]
```diff
@@ -441,6 +441,11 @@ def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
             raise ValueError(f"Received loss_type of {self.loss_type}, but the model only accepts one of ['grpo', 'dr_grpo']")
         self.scale_rewards = self.cfg.get('scale_rewards', True)
         self.max_decoder_steps = self.cfg.get('max_decoder_steps', 430)
+        # If the best record in the group is above this threshold, we will not use that group for training
+        self.best_cer_threshold = self.cfg.get('best_cer_threshold', 1.0)
+        # If the worst record in the group exceeds this threshold, we will not use that group for training
+        self.worst_cer_threshold = self.cfg.get('worst_cer_threshold', 1.0)
```
Comment (blisc) [inline nemo/collections/tts/models/magpietts_preference_optimization.py]: Why are these both defaulted to 1?

File: nemo/collections/tts/models/magpietts_preference_optimization.py [hunk 2]
```diff
@@ -636,6 +660,8 @@ def generate_and_reward(self, batch, num_generations_per_item, mode='train'):
         advantages = [x['advantage'] for x in batch_metrics]
         advantages = torch.tensor(advantages, device=self.device)
         print("Mean reward: ", all_groups_mean_reward)
+        # import ipdb; ipdb.set_trace()
```
Comment (blisc) [inline nemo/collections/tts/models/magpietts_preference_optimization.py]: Remove debugging comment

**Review by blisc — CHANGES_REQUESTED** [review]

File: nemo/collections/tts/models/magpietts_preference_optimization.py (hunk 1 above)
Comment (shehzeen) [reply nemo/collections/tts/models/magpietts_preference_optimization.py]: If it is 1.0, we don't filter out anything and include all groups in loss calculation (Since max CER is capped to 1.0)

Comment (blisc) [reply nemo/collections/tts/models/magpietts_preference_optimization.py]: Can you add that to your comment? That we cap at 1.0 and the default will not do any filtering

**Review by blisc — APPROVED** [review]

File: nemo/collections/tts/models/magpietts_preference_optimization.py (hunk 1 above)
Comment (blisc) [reply nemo/collections/tts/models/magpietts_preference_optimization.py]: I don't think we cap cer at 100% either. I've seen >1.0 on the short utterances before

Comment (shehzeen) [reply nemo/collections/tts/models/magpietts_preference_optimization.py]: Added comments.

I only do the clamping during preference optimization and not in the evaluate script. It is done at around line 556 in nemo/collections/tts/models/magpietts_preference_optimization.py

`cer_gt = min(max(cer_gt, 0.0), 1.0)  # Ensure CER is in [0, 1]
`
